text/0000-generic-dependent-consts.md

```diff
@@ -0,0 +1,291 @@
+does not know what the value in a match pattern actually is.
+
+To date we have dodged the issue by simply not allowing constant expressions to
+depend upon type parameters at all.
+
+However, in the process we have foreclosed a fairly wide range of possibilities.
+Here are several examples of functions that are forbidden, even with basic
+support for associated consts:
+
+```rust
+fn do_something_optional<T>(x: Option<T>) {
+    const NOTHING: Option<T> = None;
+    match x {
+        Some(y) => { /* ... */ }
+        NOTHING => { /* ... */ }
+    }
+}
+unsafe fn do_something_with_bitmask<T>(x: T) {
+    // Assume for the sake of argument that Sized::SIZE is an associated const
+    // of type `usize`.
+    const SIZE_IN_BITS: usize = 8 * <T as Sized>::SIZE;
+    /* ... */
+}
+fn circle_area<T: Float>(radius: T) {
+    // This redefinition is not very useful, but has a straightforward meaning.
+    const PI: T = T::PI;
+    PI*radius*radius
+}
+fn circumference<T: Float>(radius: T) {
+    const TWO_PI: T = 2.0 * T::PI;
+    TWO_PI*radius
+}
+```
+
+The first three examples would be allowed under this RFC. The fourth function,
+`circumference`, would still be disallowed because the compiler cannot verify
+during type checking that `2.0 * T::PI` is a constant expression of type
+`T`. However, this RFC removes *one* of the barriers that prevents
+implementation of this last example.
+
+Note that the question of how to handle generic constants will become even more
+pressing if user-defined types are allowed to depend on constants in the future.
+In that case, these issues will not be limited to the interaction between
+associated consts and a handful of language features.
+
+# Detailed design
+
+Because they are inlined wherever possible, non-associated constants are similar
+to macros in many circumstances. Consider this definition:
+
+```rust
+const C: T = EXPR;
+```
+
+Most uses of `C` could be replaced with `EXPR: T` (using type ascription) with
+no change in the meaning of the code. The main difference is that when a named
+constant `C` is defined, `EXPR` is required to be a constant expression, and can
+be used in certain situations (especially patterns) where inlining the
+expression, either manually or via macro, would produce invalid syntax. (Of
+course, named constants also improve readability and allow for clearer error
+messages in some cases.)
+
+Keeping this in mind, it is straightforward to understand most uses of constant
+expressions. The design below will mostly focus on special cases and
+interactions with other language features.
+
+## Other items in functions
+
+The scope of this RFC is limited to consts. Nested functions will still be
+forbidden from referencing "outer" type parameters. The same is true for
+statics. There are two justifications for this.
+
+Firstly, nested static items, whether they are static variables or functions,
+must have static locations in memory. If these items could use the type
```

Comment:
I did not expect that. [Tried it out in the playpen](http://is.gd/5ngeiK), and you're completely correct. I would have expected every parameter-combination to have its own static, but that would be impossible to get correct, since multiple crates might be using a crate with a generic function and then you'd get multiple statics for the same parameter-combination.